File: blepro/src/main/java/com/lepu/blepro/ext/er2/Er2AnalysisFile.java
package com.lepu.blepro.ext.er2;

import java.util.ArrayList;

public class Er2AnalysisFile {

    private int fileVersion;     // 文件版本 e.g.  0x01 :  V1
    private int recordingTime;   // 记录时长 e.g. 3600 :  3600s
    private ArrayList<AnalysisResult> resultList = new ArrayList<>();  // 每1分钟分析结果

    public Er2AnalysisFile(byte[] bytes) {
        com.lepu.blepro.ble.data.Er2AnalysisFile data = new com.lepu.blepro.ble.data.Er2AnalysisFile(bytes);
        fileVersion = data.getFileVersion();
        recordingTime = data.getRecordingTime();
        int len = data.getResultList().size();
        for (int i=0; i<len; i++) {
            AnalysisResult result = new AnalysisResult();
            result.result = data.getResultList().get(i).getResult();
            result.hr = data.getResultList().get(i).getHr();
            result.qrs = data.getResultList().get(i).getQrs();
            result.pvcs = data.getResultList().get(i).getPvcs();
            result.qtc = data.getResultList().get(i).getQtc();
            result.st = data.getResultList().get(i).getSt();
            resultList.add(result);
        }
    }

    public int getFileVersion() {
        return fileVersion;
    }

    public void setFileVersion(int fileVersion) {
        this.fileVersion = fileVersion;
    }

    public int getRecordingTime() {
        return recordingTime;
    }

    public void setRecordingTime(int recordingTime) {
        this.recordingTime = recordingTime;
    }

    public ArrayList<AnalysisResult> getResultList() {
        return resultList;
    }

    public void setResultList(ArrayList<AnalysisResult> resultList) {
        this.resultList = resultList;
    }

    public class AnalysisResult {
        private int result;
        private Er2EcgDiagnosis diagnosis;  // 诊断结果
        private int hr;                     // 心率 单位：bpm
        private int qrs;                    // QRS 单位：ms
        private int pvcs;                   // PVC个数
        private int qtc;                    // QTc 单位：ms
        private short st;                   // ST（以ST*100存储），单位为mV

        public int getResult() {
            return result;
        }

        public void setResult(int result) {
            this.result = result;
        }

        public Er2EcgDiagnosis getDiagnosis() {
            if (diagnosis == null) {
                diagnosis = new Er2EcgDiagnosis(result);
            }
            return diagnosis;
        }

        public void setDiagnosis(Er2EcgDiagnosis diagnosis) {
            this.diagnosis = diagnosis;
        }

        public int getHr() {
            return hr;
        }

        public void setHr(int hr) {
            this.hr = hr;
        }

        public int getQrs() {
            return qrs;
        }

        public void setQrs(int qrs) {
            this.qrs = qrs;
        }

        public int getPvcs() {
            return pvcs;
        }

        public void setPvcs(int pvcs) {
            this.pvcs = pvcs;
        }

        public int getQtc() {
            return qtc;
        }

        public void setQtc(int qtc) {
            this.qtc = qtc;
        }

        public short getSt() {
            return st;
        }

        public void setSt(short st) {
            this.st = st;
        }
    }

}

File: blepro/src/main/java/com/lepu/blepro/ble/data/Er2AnalysisFile.kt
package com.lepu.blepro.ble.data

import com.lepu.blepro.utils.ByteUtils.byte2UInt
import com.lepu.blepro.utils.ByteUtils.toSignedShort
import com.lepu.blepro.utils.bytesToHex
import com.lepu.blepro.utils.toUInt

class Er2AnalysisFile(val bytes: ByteArray) {

    var fileVersion: Int     // 文件版本 e.g.  0x01 :  V1
    // reserved 9
    var recordingTime: Int   // 记录时长 e.g. 3600 :  3600s
    // reserved 66
    var resultList = mutableListOf<AnalysisResult>()  // 每1分钟分析结果

    init {
        var index = 0
        fileVersion = byte2UInt(bytes[index])
        index++
        index += 9
        recordingTime = toUInt(bytes.copyOfRange(index, index+4))
        index += 4
        index += 66
        val len = (bytes.size - index).div(32)
        for (i in 0 until len) {
            resultList.add(AnalysisResult(bytes.copyOfRange(index+i*32, index+(i+1)*32)))
        }
    }

    override fun toString(): String {
        return """
            Er2AnalysisFile : 
            bytes : ${bytesToHex(bytes)}
            fileVersion : $fileVersion
            recordingTime : $recordingTime
            resultList : $resultList
        """.trimIndent()
    }

    class AnalysisResult(val bytes: ByteArray) {
        var result: Int
        var diagnosis: Er2EcgDiagnosis  // 诊断结果
        var hr: Int                     // 心率 单位：bpm
        var qrs: Int                    // QRS 单位：ms
        var pvcs: Int                   // PVC个数
        var qtc: Int                    // QTc 单位：ms
        var st: Short                   // ST（以ST*100存储），单位为mV
        // reserved 18
        init {
            var index = 0
            result = toUInt(bytes.copyOfRange(index, index+4))
            diagnosis = Er2EcgDiagnosis(bytes.copyOfRange(index, index+4))
            index += 4
            hr = toUInt(bytes.copyOfRange(index, index+2))
            index += 2
            qrs = toUInt(bytes.copyOfRange(index, index+2))
            index += 2
            pvcs = toUInt(bytes.copyOfRange(index, index+2))
            index += 2
            qtc = toUInt(bytes.copyOfRange(index, index+2))
            index += 2
            st = toSignedShort(bytes[index], bytes[index+1])
        }
        override fun toString(): String {
            return """
                AnalysisResult : 
                bytes : ${bytesToHex(bytes)}
                result : $result
                diagnosis : $diagnosis
                hr : $hr
                qrs : $qrs
                pvcs : $pvcs
                qtc : $qtc
                st : $st
            """.trimIndent()
        }
    }

}